lib/ruby_smb/client.rb

```diff
@@ -253,5 +262,39 @@ def wipe_state!
       self.sequence_counter = 0
       self.smb2_message_id  = 0
     end
+
+    # Parses a raw response using the supplied response packet class, which
+    # must be a subclass of {RubySMB::GenericPacket}. It raises an exception
+    # if the NTStatus is not `expected_status_code`. If an error occurs, it
+    # uses the default {RubySMB::SMB1::Packet::EmptyPacket} to parse the
+    # response.
+    #
+    # @example
+    #   raw_response = @client.send_recv(request_packet)
+    #   response = @client.parse_response(
+    #     response_packet: RubySMB::SMB1::Packet::NtCreateAndxResponse,
+    #     raw_response: raw_response)
+    #
+    # @param response_packet [Class] the class used to parse the raw response
+    # @param raw_response [String] the raw response to parse
+    # @param expected_status_code [WindowsError::ErrorCode] the expected status code
+    # @return [RubySMB::GenericPacket] the parsed response
+    # @raise [RubySMB::Error::InvalidPacket] if the response command is not the the expected command from `response_packet`
+    # @raise [RubySMB::Error::UnexpectedStatusCode] if the response NTStatus is not `expected_status_code`
+    def parse_response(response_packet:, raw_response:, expected_status_code: WindowsError::NTStatus::STATUS_SUCCESS)
```

Comment:
All this error handling seems like it might belong in the GenericPacket class. #read does some base generic handling accounting for SMB1 and SMB2. https://github.com/rapid7/ruby_smb/blob/master/lib/ruby_smb/generic_packet.rb#L42 . This seems like more packet related error handling.

Reply:
Absolutely, I removed this helper methods from the Client class and add the validation logic in the caller classes. Thanks for the heads up. Addressed in 4bb00fb.